feat: Expand LLM support with PromptModel, PromptNode, and PromptTemplate

vblagoje opened the pull request:
### Related Issues
- fixes and completes #3306 based on #3665 

### Proposed Changes:
Adds PromptModel, PromptNode, and PromptTemplate components 
Adds support to prompt the LLM model directly via PromptNode
Adds support to use PromptNode in Haystack pipelines
Adds support for T5-Flan model invocation
Adds support for OpenAI InstructGPT models

See #3665 and #3306 for more details.

### How did you test it?
Unit tests and Google Colab [notebok](https://colab.research.google.com/drive/1bOIxb8cnpTrpMtTSBArY9FJlL59Ar4K_)

### Notes for the reviewer
Component names are finalized. More details in #3665 
Default supported tasks to be expanded (see Appendix F, figure 17 of https://arxiv.org/abs/2210.11416)
Let's adhere thoroughly to #3665 and ensure community requests are honoured (see Discord discussions). 

### Checklist
- [x] I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- [x] I have updated the related issue with new insights and changes
- [x] I added tests that demonstrate the correct behavior of the change
- [x] I've used the [conventional commit convention](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title
- [x] I documented my code
- [x] I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,643 @@
+logger = logging.getLogger(__name__)
+
+
+class BaseTemplate(BaseComponent):
```
@vblagoje @ZanSara Should this be called `BasePromptTemplate` instead of `BaseTemplate` maybe?

julian-risch replied:
Otherwise users might get confused and think it has anything to do with string.Template, which we import too.

vblagoje commented on haystack/nodes/llm/prompt_node.py:
````diff
@@ -0,0 +1,643 @@
+
+    ```python
+    PromptTemplate(name="sentiment-analysis",
+                   prompt_text="Please give a sentiment for this context. Answer with positive, negative
````
I suggest we don't change this @agnieszka-m because users might then write prompts with these quotations marks 😵‍💫

vblagoje commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,643 @@
+        return template_dict
+
+
+PREDEFINED_PROMPT_TEMPLATES = [
```
This prompt goes into model not to customers :-)

vblagoje commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,643 @@
+    ),
+    PromptTemplate(
+        name="question-answering-check",
+        prompt_text="Does the following context contain the answer to the question. "
```
@agnieszka-m don't review prompts :-)

vblagoje commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,643 @@
+    ),
+    PromptTemplate(
+        name="sentiment-analysis",
+        prompt_text="Please give a sentiment for this context. Answer with positive, "
```
@agnieszka-m these prompts where taken from research papers, no need to review them

vblagoje commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,643 @@
+class OpenAIInvocationLayer(PromptModelInvocationLayer):
+    """
+    PromptModelInvocationLayer implementation for OpenAI's GPT-3 InstructGPT models.
+    See https://beta.openai.com/docs/models/gpt-3 for more details.
```
@agnieszka-m I don't get it - Do you mean to convert it to a link?

agnieszka-m replied:
exactly, make it a friendly link with a title, not just the URL.

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        self.invocation_layers[condition] = invocation_layer
+
+    def invoke(self, prompt: Union[str, List[str]], **kwargs) -> List[str]:
```
docstring with description of `prompt` parameter should be added here

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+    that the model supports. Defining additional prompt templates enables users to extend the model's capabilities
+    and use it for a broader range of NLP tasks within the Haystack ecosystem. Prompt engineers would define
+    templates for each NLP task and register them with PromptNode. The burden of defining templates for each
+    task would be on the prompt engineers, not the users.
```
@agnieszka-m I suggest removing "would" here. What do you think?

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        template for this PromptNode.
+
+        :param prompt_template: The name of the optional prompt template to use
+        :type prompt_template: Optional[Union[str, PromptTemplate]]
```
:type description is not needed here because we already have a type annotation in the method's signature.

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Adds a prompt template to the list of supported prompt templates.
+        :param prompt_template: PromptTemplate object to be added.
+        :type prompt_template: PromptTemplate
```
We just describe parameters with `:param` but `:type` can be removed.

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Removes a prompt template from the list of supported prompt templates.
+        :param prompt_template: Name of the prompt template to be removed.
+        :type prompt_template: str
```
this line can be removed

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Sets the default prompt template for the node.
+        :param prompt_template: the prompt template to be set as default.
+        :type prompt_template: Union[str, PromptTemplate]
```
this line can be removed

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Checks if a prompt template is supported.
+        :param prompt_template: the prompt template to be checked.
+        :type prompt_template: Union[str, PromptTemplate]
```
can be removed

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Returns a prompt template by name.
+        :param prompt_template_name: the name of the prompt template to be returned.
+        :type prompt_template_name: str
```
can be removed

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+        """
+        Returns the list of parameters for a prompt template.
+        :param prompt_template: the name of the prompt template.
+        :type prompt_template: str
```
can be removed

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+
+    def run(
+        self,
+        query: Optional[str] = None,
```
we need docstrings describing the parameters of this method here

julian-risch commented on pyproject.toml:
```diff
@@ -48,7 +48,7 @@ classifiers = [
 dependencies = [
   "importlib-metadata; python_version < '3.8'",
-  "torch>1.9,<1.13",
+  "torch>=1.7,!=1.12.0",
```
Is this the preferred way now that we merged https://github.com/deepset-ai/haystack/pull/3703 ? @sjrl @mayankjobanputra

julian-risch replied:
There is also a leftover in the actions of version 1.12 here: https://github.com/deepset-ai/haystack/blob/54bf7ad3439a9c481cb8d32ca720612a1341deb8/.github/actions/python_cache/action.yml#L56

sjrl replied:
Is there a reason we don't want to support version `1.12.0`? Add should the bump to Pytorch version `1.13.0` be handled in a separate PR?

sjrl replied:
> There is also a leftover in the actions of version 1.12 here:
> 
> https://github.com/deepset-ai/haystack/blob/54bf7ad3439a9c481cb8d32ca720612a1341deb8/.github/actions/python_cache/action.yml#L56

@julian-risch I've opened this PR https://github.com/deepset-ai/haystack/pull/3719 to hopefully remove the last of the references to torch-scatter.

vblagoje replied:
The main reason @sjrl is that Colab has been using 1.13.0 since a week ago. And uninstalling 1.13 and reinstalling 1.12 kills the Colab experience totally IMHO

vblagoje replied:
@julian-risch is just a temp solution that I'll remove soon and sync with the main. I had to do this because I shared PromptNode notebook demos with people.

julian-risch commented on haystack/nodes/llm/prompt_node.py:
```diff
@@ -0,0 +1,702 @@
+    ):
+        super().__init__(model_name_or_path, max_length)
+        if not isinstance(api_key, str) or len(api_key) == 0:
+            raise ValueError(
```
ValueError should be a `OpenAIError` defined in `haystack/errors.py` instead.

julian-risch commented on test/nodes/test_prompt_node.py:
```diff
@@ -0,0 +1,355 @@
+        prompt_node.prompt("invalid-task", {})
+
+
+def test_invalid_prompting(prompt_node):
```
not sure that we need test cases for this invalid input format. If you think we do, then I'd suggest we make the expected ValueError more specific.

vblagoje commented on test/nodes/test_prompt_node.py:
```diff
@@ -0,0 +1,355 @@
+    model = PromptModel()
+    assert model.model_name_or_path == "google/flan-t5-base"
+
+    with pytest.raises(ValueError):
```
Right, great - I forgot one can do this actually.

vblagoje commented:
@julian-risch @sjrl @ZanSara @mathislucka @agnieszka-m I think this one is ready to go. I addressed Aga's and Julian's comments - excellent feedback. Aside from trivial changes of `llm` to `prompt`, we now support model_kwargs for PromptModel. This is super important for model layout on GPU use-case, and we can customize it completely, even from YAML. Aside from our superusers, this will make Kristof very happy. I also added a unit test to ensure both PromptTemplate and model_kwargs work from YAML for HF and OpenAI (test_complex_pipeline_with_all_features). As we discussed, I had to make slight changes to config.py to enable this feature. Please pay close attention to the following details. Changes in config.py:
- I added model_kwargs to JSON_FIELDS as we have cases when we need to pass special parameters to the model in HF.
- I added SKIP_VALIDATION_KEYS to avoid checking prompt_text, which fails validation due to the use of $ for template variables

vblagoje commented:
@masci IMHO, this one can be safely included in 1.12. As mentioned above, it is self-contained and has one minor pipeline validation change. @ZanSara can we generate yaml schema and verify everything together?

vblagoje commented:
I also added two more default prompt templates: `language-detection` and `translation` which were tested in the updated Colab [notebook](https://colab.research.google.com/drive/1bOIxb8cnpTrpMtTSBArY9FJlL59Ar4K_)

vblagoje commented:
@bogdankostic, how can we ensure that OPENAI_API_KEY gets injected even for PRs on forks? I've tested locally on my machine with OPENAI_API_KEY, but on CI it is not injected, and I suspect it is because this PR is on my fork....

julian-risch reviewed (approved):
Looks superb to me! Let's get this going. 👍

ZanSara reviewed (commented):
Ok this was denser than I believed! Sorry for being so late in review :see_no_evil: 

First: I found NO BUGS. The code is ok. What I found was a lot of improvements, some of which I find quite important, but nothing that blocks this PR to the point that I'll request changes.

Please have a look at them. I added suggestions for all so that you don't need to write the code yourself. Let me know if some of these raise any concerns.

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+        self.prompt_text = prompt_text
+        self.prompt_params = prompt_params
+
+    def fill(self, *args, **kwargs) -> Dict[str, Any]:
+        """
+        Fills the prompt text with non-keyword and keyword arguments.
+
+        In the case of non-keyword arguments, the order of the arguments should match the left-to-right
+        order of appearance of the parameters in the prompt text. For example, if the prompt text is:
+        `Please come up with a question for the given context and the answer. Context: $documents;
+        Answer: $answers; Question:` then the first non-keyword argument will fill the $documents placeholder
+        and the second non-keyword argument will fill the $answers placeholder.
+
+        In the case of keyword arguments, the order of the arguments does not matter. Placeholders in the
+        prompt text are filled with the corresponding keyword argument.
+
+        """
+        template_dict = {}
+        # attempt to resolve args first
+        if args:
+            if len(args) != len(self.prompt_params):
+                logger.warning(
+                    f"For {self.name}, expected {self.prompt_params} arguments, instead "
+                    f"got {len(args)} arguments {args}"
+                )
+            for prompt_param, arg in zip(self.prompt_params, args):
+                template_dict[prompt_param] = [arg] if isinstance(arg, str) else arg
+        # then attempt to resolve kwargs
+        if kwargs:
+            for param in self.prompt_params:
+                if param in kwargs:
+                    template_dict[param] = kwargs[param]
+
+        if set(template_dict.keys()) != set(self.prompt_params):
+            available_params = set(list(template_dict.keys()) + list(set(kwargs.keys())))
+            raise ValueError(f"Expected prompt params {self.prompt_params} but got {list(available_params)}")
+
+        template_dict["prompt_template"] = self.prompt_text
+        return template_dict
```
@ZanSara having non-keyword arguments allows us to fill in templates by order of appearance, so we don't need to specify them explicitly with kwargs - if users choose to do so. Let's not mess with it now and allow users to use args even if most of them will use kwargs.

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+        else:
+            return self.prompt(self.default_prompt_template, *args, **kwargs)
+
+    def prompt(self, prompt_template: Optional[Union[str, PromptTemplate]], *args, **kwargs) -> List[str]:
```
We do for use cases like: `node("What is the capital of Germany?")`

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+        self.invocation_layers: Dict[Callable, Type[PromptModelInvocationLayer]] = {}
+
+        self.register(lambda x: hf_invocation_layer_supports(x), HFLocalInvocationLayer)  # pylint: disable=W0108
+        self.register(lambda x: openai_invocation_layer_supports(x), OpenAIInvocationLayer)  # pylint: disable=W0108
```
+1 good idea, much cleaner and more straightforward - see https://github.com/deepset-ai/haystack/pull/3667/commits/75c0496ffe7810eb89935fcbe56b1da37c292058

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+
+        if response.status_code != 200:
+            openai_error: OpenAIError
+            if response.status_code == 429:
+                openai_error = OpenAIRateLimitError(f"API rate limit exceeded: {response.text}")
+            else:
```
What does it mean to highlight?

ZanSara replied:
To raise a specific error for those with a clear error message

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+        elif isinstance(model_kwargs, Dict):
+            self.model_kwargs = model_kwargs
+        else:
+            self.model_kwargs = {}
```
kwargs work, great catch, thanks @ZanSara

vblagoje commented on test/nodes/test_prompt_node.py:
```diff
@@ -0,0 +1,482 @@
+    assert r[0].casefold() == "positive"
+
+
+def test_direct_prompting(prompt_node):
+    r = prompt_node("What is the capital of Germany?")
+    assert r[0].casefold() == "berlin"
+
+    r = prompt_node("What is the capital of Germany?", "What is the secret of universe?")
+    assert r[0].casefold() == "berlin"
+    assert len(r[1]) > 0
+
+    r = prompt_node("Capital of Germany is Berlin", task="question-generation")
+    assert len(r[0]) > 10 and "Germany" in r[0]
+
+    r = prompt_node(["Capital of Germany is Berlin", "Capital of France is Paris"], task="question-generation")
+    assert len(r) == 2
+
```
[fd68891](https://github.com/deepset-ai/haystack/pull/3667/commits/fd688913230629d75f33479905ec40ebd7c87440)

vblagoje commented on test/nodes/test_prompt_node.py:
```diff
@@ -0,0 +1,482 @@
+    assert "custom-task" not in prompt_node.get_prompt_template_names()
+
+
+def test_invalid_template(prompt_node):
+    with pytest.raises(ValueError, match="Invalid parameter"):
+        PromptTemplate(
+            name="custom-task", prompt_text="Custom task: $pram1 $param2", prompt_params=["param1", "param2"]
+        )
+
+    with pytest.raises(ValueError, match="Number of parameters"):
+        PromptTemplate(name="custom-task", prompt_text="Custom task: $param1", prompt_params=["param1", "param2"])
```
[fd68891](https://github.com/deepset-ai/haystack/pull/3667/commits/fd688913230629d75f33479905ec40ebd7c87440)

vblagoje commented on test/nodes/test_prompt_node.py:
```diff
@@ -0,0 +1,482 @@
+        prompt_node = PromptNode("valhalla/t5-base-e2e-qg")
+
+    with pytest.raises(ValueError, match="Model some-random-model is not supported"):
+        PromptNode("some-random-model")
```
See [fd68891](https://github.com/deepset-ai/haystack/pull/3667/commits/fd688913230629d75f33479905ec40ebd7c87440)

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+            else:
+                raise ValueError(f"{prompt_template} with args {args} , and kwargs {kwargs} not supported")
+            # we have potentially args and kwargs; task selected, so templating is needed
+            prompt_prepared = template_to_fill.fill(*args, **kwargs)
```
I know, I know, but I really like using PromptNode with non-keyword arguments....

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+                f"Please select a different name to add this prompt template."
+            )
+
+        cls.prompt_templates[prompt_template.name] = prompt_template  # type: ignore
```
This one is required: `haystack/nodes/prompt/prompt_node.py:674: error: Invalid index type "Optional[str]" for "Dict[str, PromptTemplate]"; expected type "str"`

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+    """
+
+    outgoing_edges: int = 1
+    prompt_templates: Dict[str, PromptTemplate] = {
+        prompt_template.name: prompt_template for prompt_template in PREDEFINED_PROMPT_TEMPLATES  # type: ignore
+    }
```
And this one is required: `haystack/nodes/prompt/prompt_node.py:543: error: Key expression in dictionary comprehension has incompatible type "Optional[str]"; expected type "str"`

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+                f"{prompt_template} not supported, please select one of: {self.get_prompt_template_names()}"
+            )
+
+        self.default_prompt_template = prompt_template  # type: ignore
```
+1

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+
+        return list(cls.prompt_templates[prompt_template].prompt_params)
+
+    def __eq__(self, other):
+        if isinstance(other, PromptNode):
+            if self.default_prompt_template != other.default_prompt_template:
+                return False
+            return self.model_name_or_path == other.model_name_or_path
+        return False
+
+    def __hash__(self):
+        return hash((self.default_prompt_template, self.model_name_or_path))
```
Leftover from the time when we needed to test object equality. We can keep them, no?

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+            labels=labels,
+            documents=[doc.content for doc in documents if isinstance(doc.content, str)] if documents else [],
+            **meta["invocation_context"],
+        )  # type: ignore
```
+1

vblagoje commented on haystack/pipelines/config.py:
```diff
@@ -98,7 +98,8 @@ def read_pipeline_config_from_yaml(path: Path) -> Dict[str, Any]:
 
 
-JSON_FIELDS = ["custom_query"]  # ElasticsearchDocumentStore.custom_query
+JSON_FIELDS = ["custom_query", "model_kwargs"]  # ElasticsearchDocumentStore.custom_query
```
+1

vblagoje commented on haystack/pipelines/config.py:
```diff
@@ -123,6 +124,8 @@ def validate_config_strings(pipeline_config: Any, is_value: bool = False):
                         json.loads(value)
                     except json.decoder.JSONDecodeError as e:
                         raise PipelineConfigError(f"'{pipeline_config}' does not contain valid JSON.")
+                elif key in SKIP_VALIDATION_KEYS:
+                    continue
```
What do you mean by annoying people? We can't load the template text without this skip.

ZanSara replied:
Exactly: I mean that we should rethink/remove the regex validation. At least evaluate properly what value it really brings, because it's not clear anymore.

vblagoje commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,819 @@
+        It takes a prompt and returns a list of generated text using the local Hugging Face transformers model
+        :return: A list of generated text.
+        """
+        output = None
+        if args:
+            output = self.pipe(args, max_length=self.max_length, **kwargs)
```
This is a good catch. We never actually invoke this code path.

vblagoje commented:
@ZanSara @sjrl I can merge even without your approval, but please give it another final glance before I do. I have integrated almost all of Sara's recommendations.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+        # attempt to resolve args first
+        if args:
+            if len(args) != len(self.prompt_params):
+                logger.warning(
```
I think we should continue to make sure not to use f-strings in our logging statements as we found that it can unintentionally affect performance as found in this PR https://github.com/deepset-ai/haystack/pull/3212. So instead something like 

```python
logger.warning("For %s, expected %s arguments, ...", self.name, self.prompt_params)
```
I think would be preferred.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    ),
+    PromptTemplate(
+        name="question-answering-check",
+        prompt_text="Does the following context contain the answer to the question. "
```
Should we add a question mark at the end of `Does the following context contain the answer to the question.` to be grammatically correct?

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    ),
+    PromptTemplate(
+        name="multiple-choice-question-answering",
+        prompt_text="Question:$questions ; Choose the most suitable option to answer the above question. "
```
Consider adding a space to follow the format of other prompts
```suggestion
        prompt_text="Question: $questions; Choose the most suitable option to answer the above question. "
```

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    PromptTemplate(
+        name="topic-classification",
+        prompt_text="Categories: $options; What category best describes: $documents; Answer:",
+        prompt_params=["documents", "options"],
```
Should we use `categories` as the variable name to be consistent with how the prompt is structured?

Also should `prompt_params` be updated to 
```python
prompt_params=["options", "documents"]
```
since `options` comes before `documents` in the prompt?

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    PromptTemplate(
+        name="language-detection",
+        prompt_text="Detect the language in the following context and answer with the "
+        "name of the language. Context: $documents; Answer:",
```
Are the prompt_params fields missing for this template? Or do they not need to be specified?

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    PromptTemplate(
+        name="translation",
+        prompt_text="Translate the following context to $target_language. Context: $documents; Translation:",
+    ),
```
Same for this PromptTemplate. Should we specify the prompt_params?

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    def __init__(
+        self,
+        model_name_or_path: str = "google/flan-t5-base",
+        max_length: Optional[int] = 100,
```
If the `max_length` value is optional I think the default value should be `None`. What do you think? 

If we need the default value to be set then I would recommend removing the `Optional` type.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+        model_name_or_path: str = "google/flan-t5-base",
+        max_length: Optional[int] = 100,
+        use_auth_token: Optional[Union[str, bool]] = None,
+        use_gpu: Optional[bool] = True,
```
I would remove the `Optional` type here since `initialize_device_settings` requires `use_gpu` to be a bool. And giving the `Optional` type suggests that the value `None` would be a valid value for this variable.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+            model_kwargs=model_input_kwargs,
+        )
+
+    def invoke(self, *args, **kwargs):
```
It doesn't seem like `*args` is ever used. And `PromptModel.invoke()` only passes through `**kwargs`. I think if it is not needed or used, we should consider removing it and making the new signature

```python
def invoke(self, **kwargs)
```

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    could be even remote, for example, a call to a remote API endpoint.
+    """
+
+    def __init__(self, model_name_or_path: str, max_length: Optional[int] = 100, **kwargs):
```
If the max_length value is optional I think the default value should be None. What do you think?

If we need the default value to be set then I would recommend removing the Optional type.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    """
+
+    def __init__(
+        self, api_key: str, model_name_or_path: str = "text-davinci-003", max_length: Optional[int] = 100, **kwargs
```
If the max_length value is optional I think the default value should be None. What do you think?

If we need the default value to be set then I would recommend removing the Optional type.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+    def __init__(
+        self,
+        model_name_or_path: str = "google/flan-t5-base",
+        max_length: Optional[int] = 100,
```
If the max_length value is optional I think the default value should be None. What do you think?

If we need the default value to be set then I would recommend removing the Optional type.

sjrl commented on haystack/nodes/prompt/prompt_node.py:
```diff
@@ -0,0 +1,822 @@
+        model_name_or_path: Union[str, PromptModel] = "google/flan-t5-base",
+        default_prompt_template: Optional[Union[str, PromptTemplate]] = None,
+        output_variable: Optional[str] = None,
+        max_length: Optional[int] = 100,
```
If the max_length value is optional I think the default value should be None. What do you think?

If we need the default value to be set then I would recommend removing the Optional type.
